private/bufpkg/bufwasm/bufwasm.go

```diff
@@ -0,0 +1,255 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufwasm
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"testing/fstest"
+
+	wasmpluginv1 "github.com/bufbuild/buf/private/gen/proto/go/buf/alpha/wasmplugin/v1"
+	"github.com/bufbuild/buf/private/pkg/protoencoding"
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/api"
+	gojs "github.com/tetratelabs/wazero/imports/go"
+	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
+	"github.com/tetratelabs/wazero/sys"
+	"go.uber.org/multierr"
+	"google.golang.org/protobuf/encoding/protowire"
+	"google.golang.org/protobuf/proto"
+)
+
+// CustomSectionName is the name of the custom wasm section we look into for buf
+// extensions.
+const CustomSectionName = ".bufplugin"
+
+// https://www.w3.org/TR/2019/REC-wasm-core-1-20191205/#sections%E2%91%A0
+const customSectionID = 0
+
+// EncodeBufSection encodes the pluginmetadata message as a custom wasm section.
+// The resulting bytes can be appended to any valid wasm file to add the new
+// section to that file.
+func EncodeBufSection(metadata *wasmpluginv1.Metadata) ([]byte, error) {
```

Comment:
global functions should go below structs/interfaces, see "Go Code Standards" in notion, section "Within a file"